### ISSUE TITLE
Keras "missing from loss dictionary"

### ISSUE THREAD

**Issue by bhack** [tensorflow/tensorflow#16493]
I think that this warning can be rephrased:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/_impl/keras/engine/training.py#L640-L641

Some outputs could be consumed as TB summary and not involved in any loss. 
Especially when you use an estimator converted from a tf.keras model  you have not the explicit control of the model_fn for placing summaries. 

So I suppose that one of the entry point for connecting summaries to the graph is the tf.train.SessionRunHook.

I.e. you can produce the output from the Dataset api, consume it as TB summary in the SessionRunHook without using it in any loss.

See also https://github.com/tensorflow/tensorflow/issues/14879#issuecomment-351996902 

/cc @fchollet

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 100 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 116 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 131 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 148 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by bhack** [comment]
@ewilderj As you can see for something quite trivial it is impossibile to understand why a triage is missing after 148 days. I cannot reverse engineering the status of this issue.

**Comment by pavithrasv** [comment]
@bhack What do you think the warning should be and can you if possible share details and code snippet of how you are using the output that does not have a corresponding loss (may be the code you use to log summaries) ? These details would help us understand your use case better and improve the code accordingly.

Thank you,
Pavithra

**Comment by bhack** [comment]
@pavithrasv  The use case it is quite simple. Suppose that one of your input from the Dataset api is just a reference index batch. You could consume in a text summary but of course not in the loss. So I think that the check and warning need to works only when the input is not consumed by anything. One endpoint could be the loss but also a summary is a valid endpoint. It need to be the same use case as the unsued variable in a linter.

**Comment by pavithrasv** [comment]
Thank you for the explanation @bhack.

Do you train using `fit` or some other way? Can you please share code snippet for how you use the additional output with summaries? I am asking because we currently do one to one mapping between the loss+metrics labels and the outputs obtained from the session as seen here:
https://github.com/keras-team/keras/blob/master/keras/engine/training_arrays.py#L156. This would mean that any output that does not correspond to a loss/metric will be dropped.

**Comment by bhack** [comment]
No I was using `model_to_estimator` and `train_and_evaluate` path.

**Comment by bhack** [comment]
@pavithrasv I don't know if is clear. The problem is in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training.py#L211. 
We cannot only check that  `if name not in loss:` and notify:
```
"and we will not be expecting '
'any data to be passed to "' + name + '" during training.'"
```
Cause this is not true and misleading. An output could be simply involved in a `tf.summary` only. So the data will pass from inputs to outputs for every batch but these outputs are not involved in a loss like "normal" outputs and input labes.

**Comment by pavithrasv** [comment]
@bhack How do you get a handle to the output that is not used in loss when using model_to_estimator? 
Here is what I have tried: https://gist.github.com/pavithrasv/2077a1e4be619cc000d573180b863ea3

In the above snippet, `dense_1` is the output that is not being used in loss. But I do not have a way to get a handle to that output from the estimator instance. 

As I have mentioned before it is difficult for us to be able to reproduce the issue without a code snippet. It would be great if you can share your code snippet.

Thank you!

**Comment by bhack** [comment]
@pavithrasv I've commented your gist directly.
I understand that consuming dense_1 is not so easy and everybody want a simpler method to consume these kind of things with `model_to_estimator` but this is what we have now with this kind of high level api.

I want to try to support you but the bug was opened on Jan so I could be a little bit hard to me to invest time to extract snippets as I had in Jan/Feb/Marc cause I am on another activity now.

I hope that we can solve this interacting on your gist especially if the code will be ready to copy'n'paste in a colab notebook.

**Comment by pavithrasv** [comment]
Changing the warning message to 'Output "' + name + '" missing from loss dictionary. We assume this was done on purpose. The fit and evaluate APIs will not be expecting any data to be passed to "' + name + '".'
